contracts/rosca.sol

```diff
@@ -1,266 +1,258 @@
-            members[msg.sender].allowance = 0;                      // send() completes its task
-
-
-            if (!opt_destination.send(amountToWithdraw)) {   //if the send() fails, put the allowance back to its original place
-                // No need to call throw here, just reset the amount owing
-                members[msg.sender].allowance = amountToWithdraw;
-                return false;
-            //TODO(shine) : look into security about withdraw
-            }
-            else{
-                LogFundsWithdrawal(msg.sender, amountToWithdraw, opt_destination);
-                return true;
-            }
-        }
-        else
-            throw;
-
-    }
-
-    function integrityCheck(uint option) // option will specify whether it will be a message call or transaction call to freeze the contract
+    else throw;
+  }
+
+  // v2: function leave();
+
+  /**
+   * Processes a periodic contribution from msg.sender ().
+   * Any excess funds will be withdrawable through withdraw().
+   */
+  function contribute() payable
+  {
+    //Product condsideration : are we gonna limit how much ether could be sent? is it total, or how much sent each time?
+    if(members[msg.sender].alive)
     {
-        /* Checklist for integrityCheck
-        *
-        */
-
+      members[msg.sender].contributed = msg.value;
+      //TODO(shine) : look into security about deposit and withdrawal
+      LogContributionMade(msg.sender, msg.value);
     }
-    // TODO(ron): escape hatch.
-
-    /**
-     * Vouches that uniqName is now using newWallet as its primary account.
-     * If X-out-of-Y people vouch, the address is updated after a timelock,
-     * unless cancelRestoreAccess() is called from the old address.
-     */
-    function restoreAccess(string uniqName, address newWallet){
-        throw; //for now
+    else
+      throw;
+  }
+
+  /**
+   * Registers a bid from msg.sender. If msg.sender has already won a round or bid is higher than lowestBid,
+   * this method will throw.
+   */
+  function bid(uint distrubtionAmountInWei)
+  {
+    if(distrubtionAmountInWei < lowestBid || !(members[msg.sender].paid < currentEpoch))
+    {
+      //set new lowestBid and winnerAddress, also trigger event new Bid placed
+      lowestBid = distrubtionAmountInWei;
+      winnerAddress = msg.sender;
+      LogNewLowestBid(lowestBid, winnerAddress);
+    }else
+      throw;
+  }
+
+  /**
+   * Withdraws available funds for msg.sender. If opt_destination is specified,
+   * sends the fund to that address.
+   */
+  function withdraw(address opt_destination)
+    returns(bool success)
+  {
+    if(members[msg.sender].alive && members[msg.sender].pendingWithdrawl > 0)
+    {
+      uint amountToWithdraw = members[msg.sender].pendingWithdrawl;  //use a temporary variable to avoid the receiver calling the withdraw function again before
+      members[msg.sender].pendingWithdrawl = 0;                      // send() completes its task
```

Comment:
since opt_desitination is optional, i think we need to add:
if (opt_destination == 0)
  opt_destination = msg.sender;